### PR TITLE
add Gatekeeper validation to manifests

### PR DESCRIPTION
**Pull request by OliverMKing**
# Description

This PR adds validation to our managed resources to ensure they are following Kubernetes best practices. We accomplish this by scanning our fixtures which represent the resources that we manage. Unfortunately, there's no tool that scans manifests with a complete library of best practices built in. We had a few options
- SafeGuards. Would be great to use since we (AKS) maintain this and can collaborate with SafeGuards team. Will likely switch to this sometime in the future. Security best practices and other ones are being added.
- [ShieldGuard](https://github.com/Azure/ShieldGuard). Also maintained by AKS. Doesn't have any best practices built into the OSS version so we'd have to maintain our own fully. This also doesn't export their Go code so we can't call it within our unit tests which is a negative.
- Gatekeeper / Gator. Exports code so we can call them through unit tests. We have to define some parts of how best practices apply but they have a [library](https://github.com/open-policy-agent/gatekeeper-library) we pull from so we get most things out of the box. 

We chose Gatekeeper for now. See the [doc](https://github.com/Azure/aks-app-routing-operator/blob/2ca61b79d12d3b61195aeb3731fce20dc3671dcd/pkg/manifests/policy/README.md) added to this PR for information on what we pull automatically and what we have to define.

Along with implementing these tests there were a few other things we had to do
1. Swap our fixtures from json to yaml. Gatekeeper / Gator doesn't handle json manifests stored in the same file. Yaml is easier to read anyways
2. Update a few dependencies (and make super minor changes to handle any breaking changes based on these bumps)

When reviewing this PR note that the files in `pkg/manifests/policy/manifests/templates` are pulled directly from the Gatekeeper Library with no changes. It may be good to view those separately so you can digest the contents of this PR better. This is a large PR but the majority of the diff is due to autogenerated things like fixtures switching from json to yaml and the autopulled templates.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

This is a testing improvement.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/manifests/fixtures/common/another-namespace.json [hunk 1]
```diff
@@ -1,15 +1,10 @@
...
-        "labels": {
-          "openservicemesh.io/monitored-by": "osm"
-        }
-      },
-      "spec": {},
-      "status": {}
-    }
-  ]
\ No newline at end of file
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (OliverMKing) [reply pkg/manifests/fixtures/common/another-namespace.json:1]: https://raw.githubusercontent.com/Azure/aks-app-routing-operator/d345f5de7008509d5c963f1f4db78008531b3dd5/pkg/manifests/fixtures/common/namespace.json

this is just a github syntax highlighter bug

[the remark this answers is not in the document]
Comment (bfoley13) [reply pkg/manifests/fixtures/common/another-namespace.json:1]: you still have the file named as .json but are using yaml

[the remark this answers is not in the document]
Comment (OliverMKing) [reply pkg/manifests/fixtures/common/another-namespace.json]: that explains that one. fixed!

File: pkg/manifests/policy/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+- Constraints. Define how the rule will actually apply to manifests. We have to manually define these by implementing the ConstraintTemplates. After pulling any new ConstraintTemplates, if they're useful to our needs, we should implement them with a Constraint. Not all ConstraintTemplates should be implemented since some are more for organizational policy rather than generic best practices.
+
+We need to periodically update the templates to pull any new best practices using the command above. The process should look like
+1. Run the above kustomize command to pull any new ConstraintTemplates
+2. Check the git diff. If no new Templates were added we can stop.
+3. Look at the new Templates and decide if they would be useful to us. Are they related to best practices and security or are they more for organizational policy? If it's in the first category add a Constraint to the manifests directory implementing the Template.
```
[the remark this answers is not in the document]
Comment (OliverMKing) [reply pkg/manifests/policy/README.md:12]: it's super doable to automate the PR creation updating the templates

**Review by bfoley13 — APPROVED** [review]

**Pull request merged**